### ISSUE TITLE
No indication when invalid/unreachable AD server

### ISSUE THREAD

**Issue by jadedfire** [gheeres/node-activedirectory#113]
In order to call AD for authentication in a way that MS supports, one must iterate through DNS SRV records and attempt to connect to each server until one is found that is both active and responsive. Using a virtual name, such as ldap://ldap.somedomain.com:1234, that resolves to a load balancer is not supported by MS (nor is it supported in my company). Thus, to connect to AD for authentication, one must sit in a for() loop attempting to connect, and perform a request that responds with either a success or failure. This results in 1 of 3 states: unable to connect; failure to autenticate; or successful authentication. Since node is asynchronous, one would expect a failure to connect, or even to resolve the server name, to come in through the callback within the error object, or through an exception. However, when an AD server is unreachable or no longer valid, the callback is never called, no exception is thrown, nor are there any valid values returned from the authenticate() function.

**Comment by gheeres** [comment]
I just added mocha tests for the scenario you described:

```
  ActiveDirectory
    #authenticate()
      ✓ should return err (ENOTFOUND) on invalid hostname (dns)
      ✓ should return err (ECONNREFUSED) on non listening port
```

Please amend or update the tests for the condition(s) that you are encountering.

**Comment by jadedfire** [comment]
I don't have time to set up a repo right now for a pull request, but you could add this:
```
    it('should return err (ETIMEDOUT) on unresponsive server', function(done) {
      var ad = new ActiveDirectory({
        url: 'ldap://[NEED URL FOR LDAP SERVER THAT TIMES OUT OR AN EMULATOR]/'
      });
      ad.authenticate(username, password, function(err, auth) {
        assert((err || {}).code === 'ETIMEDOUT');
        assert(! auth);
        done();
      });
    });
```
NOTE: that test will require either an emulator or LDAP (AD) server that is hung, or too busy to reply. Perhaps a node socket server listening on port 389 that only handles the connect and then spins??? I should also add that my DEV environment is on Windows, in case this happens to be an OS specific issue.

**Comment by gheeres** [comment]
By default the 'timeout' defaults to infinity. This is an ldapjs option. 

```
timeout	Milliseconds client should let operations live for before timing out (Default: Infinity)
```

Please try specifying a timeout option inside of your config object:

```javascript
  var ad =  new ActiveDirectory({
    url: 'ldap://domain.com',
    timeout: 1000
  });
```

See the [ldapjs client documentation](http://ldapjs.org/client.html) for additional options that can be passed through to the underlying ldapjs implementation.

Otherwise, I'm unable to simulate an ETIMEDOUT event. Here is what I put together to try to simulate a timed out / failed request, but its not testing the right condition(s):

```javascript
    it('should return err (ETIMEDOUT) on unresponsive server', function(done) {
      var timeout = 6000;
      var port = 3890;

      this.timeout(timeout);

      var server = net.createServer(function(socket) {
        socket.setKeepAlive(false, 10);
        socket.setTimeout(10, function() {
          socket.end();
        });
      });
      server.listen(port, '127.0.0.1');
      var ad = new ActiveDirectory({
        url: 'ldap://127.0.0.1:'+port+'/',
        timeout: timeout - 1000
      });
      ad.authenticate(username, password, function(err, auth) {
        assert((err || {}).code === 'ETIMEDOUT');
        assert(! auth);

        server.close();
        done();
      });
    });
```

I'm getting a CLOSED / ConnectionError with the above code instead of a time out error.

```
{ [ConnectionError: 1__ldap://127.0.0.1:3890/ closed]
  dn: [Getter],
  code: [Getter],
  name: [Getter],
  message: [Getter] }
``` 

Aside from using iptables (Linux) to DISCARD packets, I'm unaware of how to simulate a timed out or unresponsive connection from within node.js.

**Comment by jadedfire** [comment]
Out of curiosity, are the two tests you added, ENOTFOUND and ECONNREFUSED, passing when you run them? They fail on my Windows machine:
```
ActiveDirectory
    #authenticate()
      1) should return err (ENOTFOUND) on invalid hostname (dns)
      2) should return err (ECONNREFUSED) on non listening port


  0 passing (4s)
  2 failing

  1) ActiveDirectory #authenticate() should return err (ENOTFOUND) on invalid hostname (dns):
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
      at null.<anonymous> ([PATH omitted for brevity]\node_modules\mocha\lib\runnable.js:189:19)

  2) ActiveDirectory #authenticate() should return err (ECONNREFUSED) on non listening port:
     Error: timeout of 2000ms exceeded. Ensure the done() callback is being called in this test.
      at null.<anonymous> ([PATH omitted for brevity]\node_modules\mocha\lib\runnable.js:189:19)
```
This is indicative of the issue I am seeing, and why I started this thread. As stated, please keep in mind that I am using Windows. If your tests are passing on Linux, then I suspect this may be OS specific.

**Comment by gheeres** [comment]
Yes, both tests are passing. Running on Linux.  I'll see if I can replicate on a windows machine.

**Comment by gheeres** [comment]
I was able to replicate the problem on Windows. It appears that Windows handles system level errors different than Linux. i.e. The ldapjs .bind() callback is not called when the .on('error') event handler is defined. Removing the .on('error') event handler results in an unhandled exception on Windows but calls the callback on Linux. To support the difference in error handling, I've added code to "prevent" the callback from getting called more than once. Since my Linux environment was still running 0.12.2, I even updated to 5.0.0 thinking that something fundamental had changed but I got the same results.

Please pull and test the branch [issue-113](https://github.com/gheeres/node-activedirectory/tree/issue-113) that I created to see if that resolves your problem. If it does, I'll merge the branch and publish a new release.

**Comment by jadedfire** [comment]
Confirmed the fix on Windows:
```
  ActiveDirectory
    #authenticate()
      √ should return err (ENOTFOUND) on invalid hostname (dns)
      √ should return err (ECONNREFUSED) on non listening port (1042ms)


  2 passing (1s)
```

**Comment by gheeres** [comment]
Merged v0.7.2
